File: backend/src/modules/monitoring/monitoring.service.ts
import { Injectable } from '@nestjs/common';
import { KamailioRpcService } from '../../common/services/kamailio-rpc.service';
import { SubscriberService } from '../subscriber/subscriber.service';
import { LocationService } from '../location/location.service';
import { AccService } from '../acc/acc.service';

@Injectable()
export class MonitoringService {
  constructor(
    private kamailioRpcService: KamailioRpcService,
    private subscriberService: SubscriberService,
    private locationService: LocationService,
    private accService: AccService,
  ) {}

  /**
   * 检查 Kamailio 服务状态
   */
  async checkHealth(): Promise<{
    status: 'online' | 'offline';
    uptime?: any;
    version?: string;
  }> {
    try {
      const isOnline = await this.kamailioRpcService.ping();
      if (!isOnline) {
        return { status: 'offline' };
      }

      const [uptime, coreInfo] = await Promise.all([
        this.kamailioRpcService.getUptime(),
        this.kamailioRpcService.getCoreInfo(),
      ]);

      return {
        status: 'online',
        uptime,
        version: coreInfo?.version,
      };
    } catch (error) {
      return { status: 'offline' };
    }
  }

  /**
   * 获取仪表盘数据
   */
  async getDashboardData(): Promise<{
    subscriberCount: number;
    onlineCount: number;
    todayCalls: number;
    todayMissedCalls: number;
    successRate: number;
    kamailioStatus: 'online' | 'offline';
  }> {
    const [subscriberStats, onlineCount, todayStats, health] = await Promise.all([
      this.subscriberService.getStats(),
      this.locationService.getOnlineCount(),
      this.accService.getTodayStats(),
      this.checkHealth(),
    ]);

    return {
      subscriberCount: subscriberStats.total,
      onlineCount,
      todayCalls: todayStats.totalCalls,
      todayMissedCalls: todayStats.missedCalls,
      successRate: todayStats.successRate,
      kamailioStatus: health.status,
    };
  }

  /**
   * 获取 Kamailio 统计信息
   */
  async getStatistics(group?: string): Promise<any> {
    return this.kamailioRpcService.getStatistics(group);
  }

  /**
   * 获取核心信息
   */
  async getCoreInfo(): Promise<any> {
    return this.kamailioRpcService.getCoreInfo();
  }

  /**
   * 获取活动对话
   */
  async getActiveDialogs(): Promise<{
    count: number;
    dialogs: any[];
  }> {
    try {
      const [count, dialogs] = await Promise.all([
        this.kamailioRpcService.getDialogCount(),
        this.kamailioRpcService.getDialogList(),
      ]);

      return {
        count: count?.active || 0,
        dialogs: dialogs || [],
      };
    } catch (error) {
      return { count: 0, dialogs: [] };
    }
  }

  /**
   * 检查 Kamailio 关键模块是否可用
   */
  async checkModules(): Promise<{
    connected: boolean;
    version?: string;
    modules: { name: string; available: boolean; error?: string }[];
  }> {
    const modules: { name: string; rpc: string; params?: any[] }[] = [
      { name: 'core', rpc: 'core.version' },
      { name: 'usrloc', rpc: 'ul.dump', params: [] },
      { name: 'dispatcher', rpc: 'dispatcher.list' },
      { name: 'drouting', rpc: 'drouting.gw_status' },
      { name: 'domain', rpc: 'domain.dump' },
      { name: 'permissions', rpc: 'permissions.addressDump' },
      { name: 'dialog', rpc: 'dlg.stats_active' },
      { name: 'cfg_rpc', rpc: 'cfg.list' },
      { name: 'statistics', rpc: 'stats.get_statistics', params: ['all'] },
      { name: 'tls', rpc: 'tls.list' },
      { name: 'pike', rpc: 'pike.list' },
    ];

    const settled = await Promise.allSettled(
      modules.map((mod) =>
        this.kamailioRpcService.call(mod.rpc, mod.params || []),
      ),
    );

    const results = modules.map((mod, i) => {
      const s = settled[i];
      if (s.status === 'fulfilled') {
        return { name: mod.name, available: true };
      }
      return {
        name: mod.name,
        available: false,
        error: s.reason instanceof Error ? s.reason.message : 'Unknown error',
      };
    });

    const coreIdx = modules.findIndex((m) => m.name === 'core');
    const connected = coreIdx >= 0 && settled[coreIdx].status === 'fulfilled';
    let version: string | undefined;
    if (coreIdx >= 0 && settled[coreIdx].status === 'fulfilled') {
      const v = (settled[coreIdx] as PromiseFulfilledResult<unknown>).value;
      version = typeof v === 'string' ? v : undefined;
    }

    return { connected, version, modules: results };
  }

  /**
   * 获取系统概览
   */
  async getSystemOverview(): Promise<{
    kamailio: {
      status: 'online' | 'offline';
      version?: string;
      uptime?: any;
    };
    database: {
      subscribers: number;
      domains: number;
      dispatchers: number;
    };
    realtime: {
      onlineUsers: number;
      activeDialogs: number;
    };
  }> {
    const health = await this.checkHealth();
    const subscriberStats = await this.subscriberService.getStats();
    const onlineCount = await this.locationService.getOnlineCount();
    
    let activeDialogs = 0;
    try {
      const dialogCount = await this.kamailioRpcService.getDialogCount();
      activeDialogs = dialogCount?.active || 0;
    } catch (error) {
      // 忽略错误
    }

    return {
      kamailio: {
        status: health.status,
        version: health.version,
        uptime: health.uptime,
      },
      database: {
        subscribers: subscriberStats.total,
        domains: subscriberStats.domains.length,
        dispatchers: 0, // 需要注入 DispatcherService
      },
      realtime: {
        onlineUsers: onlineCount,
        activeDialogs,
      },
    };
  }
}

File: kam-admin-module-monitor/src/main/java/io/kamailio/admin/modules/monitoring/service/MonitoringService.java
package io.kamailio.admin.modules.monitoring.service;

import io.kamailio.admin.common.RpcTimeouts;
import io.kamailio.admin.common.service.KamailioRpcService;
import io.kamailio.admin.modules.acc.service.AccService;
import io.kamailio.admin.modules.dispatcher.service.DispatcherService;
import io.kamailio.admin.modules.location.service.LocationService;
import io.kamailio.admin.modules.subscriber.service.SubscriberService;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Service
public class MonitoringService {
    private final KamailioRpcService kamailioRpc;
    private final SubscriberService subscriberService;
    private final LocationService locationService;
    private final AccService accService;
    private final DispatcherService dispatcherService;

    public MonitoringService(KamailioRpcService kamailioRpc, SubscriberService subscriberService,
                            LocationService locationService, AccService accService,
                            DispatcherService dispatcherService) {
        this.kamailioRpc = kamailioRpc;
        this.subscriberService = subscriberService;
        this.locationService = locationService;
        this.accService = accService;
        this.dispatcherService = dispatcherService;
    }

    public Map<String, Object> checkHealth() {
        if (Boolean.FALSE.equals(kamailioRpc.ping().block(RpcTimeouts.DEFAULT_BLOCK)))
            return Map.of("status", "offline");
        try {
            Object uptime = kamailioRpc.getUptime().block(RpcTimeouts.DEFAULT_BLOCK);
            Object coreInfo = kamailioRpc.getCoreInfo().block(RpcTimeouts.DEFAULT_BLOCK);
            String version = coreInfo instanceof Map ? (String) ((Map<?, ?>) coreInfo).get("version") : null;
            return Map.of("status", "online", "uptime", uptime != null ? uptime : "", "version", version != null ? version : "");
        } catch (Exception e) {
            return Map.of("status", "offline");
        }
    }

    public Map<String, Object> getDashboardData() {
        var subscriberStats = subscriberService.getStats();
        long onlineCount = locationService.getOnlineCount();
        var todayStats = accService.getTodayStats();
        var health = checkHealth();
        return Map.of(
                "subscriberCount", subscriberStats.get("total"),
                "onlineCount", onlineCount,
                "todayCalls", todayStats.get("totalCalls"),
                "todayMissedCalls", todayStats.get("missedCalls"),
                "successRate", todayStats.get("successRate"),
                "kamailioStatus", health.get("status")
        );
    }

    public Object getStatistics(String group) {
        return kamailioRpc.getStatistics(group).block(RpcTimeouts.DEFAULT_BLOCK);
    }

    public Object getCoreInfo() {
        return kamailioRpc.getCoreInfo().block(RpcTimeouts.DEFAULT_BLOCK);
    }

    public Map<String, Object> getActiveDialogs() {
        try {
            Object count = kamailioRpc.getDialogCount().block(RpcTimeouts.DEFAULT_BLOCK);
            Object dialogs = kamailioRpc.getDialogList().block(RpcTimeouts.DEFAULT_BLOCK);
            int c = 0;
            if (count instanceof Map<?, ?> m) { Object v = m.get("active"); if (v instanceof Number n) c = n.intValue(); }
            return Map.of("count", c, "dialogs", dialogs != null ? dialogs : List.of());
        } catch (Exception e) {
            return Map.of("count", 0, "dialogs", List.of());
        }
    }

    public Map<String, Object> checkModules() {
        var mods = List.of("core", "usrloc", "dispatcher", "drouting", "domain", "permissions", "dialog", "cfg_rpc", "statistics", "tls", "pike");
        var rpcs = List.of("core.version", "ul.dump", "dispatcher.list", "drouting.gw_status", "domain.dump", "permissions.addressDump", "dlg.stats_active", "cfg.list", "stats.get_statistics", "tls.list", "pike.list");
        var results = new ArrayList<Map<String, Object>>();
        boolean connected = false;
        String version = null;
        for (int i = 0; i < mods.size(); i++) {
            try {
                Object r = kamailioRpc.call(rpcs.get(i), rpcs.get(i).equals("stats.get_statistics") ? new Object[]{"all"} : new Object[]{}).block(RpcTimeouts.DEFAULT_BLOCK);
                if (i == 0 && r != null) { connected = true; version = r.toString(); }
                results.add(Map.of("name", mods.get(i), "available", true));
            } catch (Exception e) {
                results.add(Map.of("name", mods.get(i), "available", false, "error", e.getMessage()));
            }
        }
        return Map.of("connected", connected, "version", version != null ? version : "", "modules", results);
    }

    public Map<String, Object> getSystemOverview() {
        var health = checkHealth();
        var subscriberStats = subscriberService.getStats();
        long onlineCount = locationService.getOnlineCount();
        long dispatcherCount = ((Number) dispatcherService.getStats().get("total")).longValue();
        int activeDialogs = 0;
        try {
            Object dc = kamailioRpc.getDialogCount().block(RpcTimeouts.DEFAULT_BLOCK);
            if (dc instanceof Map<?, ?> m) { Object v = m.get("active"); if (v instanceof Number n) activeDialogs = n.intValue(); }
        } catch (Exception ignored) {}
        return Map.of(
                "kamailio", Map.of("status", health.get("status"), "version", health.get("version"), "uptime", health.get("uptime")),
                "database", Map.of("subscribers", subscriberStats.get("total"), "domains", ((List<?>) subscriberStats.get("domains")).size(), "dispatchers", dispatcherCount),
                "realtime", Map.of("onlineUsers", onlineCount, "activeDialogs", activeDialogs)
        );
    }
}
